diff -Nur mysql.1/mysql-4.1.25/mysys/default.c mysql-4.1.25.osv/mysys/default.c
--- mysql.1/mysql-4.1.25/mysys/default.c	2008-12-30 20:15:05.000000000 +0900
+++ mysql-4.1.25.osv/mysys/default.c	2014-06-04 08:10:38.405237838 +0900
@@ -420,6 +420,7 @@
     strmov(name,config_file);
   }
   fn_format(name,name,"","",4);
+#if 0
 #if !defined(__WIN__) && !defined(OS2) && !defined(__NETWARE__)
   {
     MY_STAT stat_info;
@@ -439,6 +440,7 @@
     }
   }
 #endif
+#endif
   if (!(fp= my_fopen(fn_format(name, name, "", "", 4), O_RDONLY, MYF(0))))
     return 0;					/* Ignore wrong files */
 
diff -Nur mysql.1/mysql-4.1.25/sql/mysqld.cc mysql-4.1.25.osv/sql/mysqld.cc
--- mysql.1/mysql-4.1.25/sql/mysqld.cc	2008-12-30 20:17:00.000000000 +0900
+++ mysql-4.1.25.osv/sql/mysqld.cc	2014-06-04 09:47:00.735482923 +0900
@@ -1389,6 +1391,7 @@
     }
   }
 
+#if 0
   if ((user_info= check_user(mysqld_user)))
   {
 #if defined(HAVE_MLOCKALL) && defined(MCL_CURRENT)
@@ -1398,6 +1401,7 @@
 #endif
       set_user(mysqld_user, user_info);
   }
+#endif
 
 #ifdef __NT__
   /* create named pipe */


