--- upstream/ffmpeg-4.0.2/fftools/ffmpeg.c	2018-07-18 13:52:07.000000000 +0000
+++ ffmpeg.c	2018-11-19 18:56:28.434006165 +0000
@@ -4849,6 +4849,6 @@
     if ((decode_error_stat[0] + decode_error_stat[1]) * max_error_rate < decode_error_stat[1])
         exit_program(69);
 
-    exit_program(received_nb_signals ? 255 : main_return_code);
+    ffmpeg_cleanup(received_nb_signals ? 255 : main_return_code);
     return main_return_code;
 }
